In [2]:
import tensorflow as tf

In [3]:
import numpy as np

In [3]:
print(tf.__version__)

2.10.1


In [4]:
random_1 = tf.random.Generator.from_seed(42)
random_1 = random_1.normal(shape=(3,2))

In [5]:
random_1

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [19]:
random_2 = tf.random.Generator.from_seed(42)
random_2 = random_2.normal(shape =(3,2))
random_2 = tf.random.shuffle(random_2, seed=20)

In [20]:
random_2

<tf.Tensor: shape=(3, 2), dtype=float32, numpy=
array([[-0.7565803 , -0.06854702],
       [ 0.07595026, -1.2573844 ],
       [-0.23193765, -1.8107855 ]], dtype=float32)>

In [11]:
random_1 == random_2

<tf.Tensor: shape=(3, 2), dtype=bool, numpy=
array([[ True,  True],
       [ True,  True],
       [ True,  True]])>

Operations that rely on a random seed actually derive it from two seeds: the global and operation-level seeds. This sets the global seed.
https://www.tensorflow.org/api_docs/python/tf/random/set_seed

In [24]:
not_shuffled = tf.constant([[10,7],
                            [3,4],
                            [2,5]
                           ])

In [25]:
not_shuffled

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]])>

In [60]:
tf.random.shuffle(not_shuffled, seed=42)
# if use only this,though seed has been set, but result still random

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 2,  5],
       [ 3,  4],
       [10,  7]])>

In [70]:
tf.random.set_seed(42)    # global level random seed
tf.random.shuffle(not_shuffled, seed=42)    # operation level random seed

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4],
       [ 2,  5]])>

In [13]:
n = np.ones((2,2))

In [14]:
type(n)

numpy.ndarray

In [16]:
a = tf.ones([3,2,3,4])

In [17]:
type(a)

tensorflow.python.framework.ops.EagerTensor

In [87]:
a

<tf.Tensor: shape=(3, 2, 3, 4), dtype=float32, numpy=
array([[[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]],


       [[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]],


       [[[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]],

        [[1., 1., 1., 1.],
         [1., 1., 1., 1.],
         [1., 1., 1., 1.]]]], dtype=float32)>

In [88]:
a.ndim

4

In [19]:
tf.ones((2,2)) # bracket

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[1., 1.],
       [1., 1.]], dtype=float32)>

In [90]:
tf.ones([2,2])   # square bracket

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[1., 1.],
       [1., 1.]], dtype=float32)>

In [91]:
tf.zeros(shape=(3,4))   #bracket

<tf.Tensor: shape=(3, 4), dtype=float32, numpy=
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]], dtype=float32)>

In [20]:
tf.zeros([2,2])

<tf.Tensor: shape=(2, 2), dtype=float32, numpy=
array([[0., 0.],
       [0., 0.]], dtype=float32)>

In [95]:
b = tf.ones(shape=(3, 2))

In [96]:
b.ndim

2

In [8]:
import numpy as np
numpy_A = np.arange(1,25, dtype= np.int32)

# X = tf.constant(some_matrix) # capital for matrix or tensor
# Y = tf.constant(vector) # non-capital for vector


In [9]:
numpy_A

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24])

In [10]:
A =tf.constant(numpy_A, shape=(2,3,4))
B =tf.constant(numpy_A)

A,B

(<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
 array([[[ 1,  2,  3,  4],
         [ 5,  6,  7,  8],
         [ 9, 10, 11, 12]],
 
        [[13, 14, 15, 16],
         [17, 18, 19, 20],
         [21, 22, 23, 24]]])>,
 <tf.Tensor: shape=(24,), dtype=int32, numpy=
 array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23, 24])>)

In [ ]:
tf.zeros(shape=(2,3), dtype= tf.float16)

In [22]:
a = tf.zeros([2, 3, 4, 5])
a

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [37]:
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]])>

In [35]:
A[:1]

<tf.Tensor: shape=(1, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]]])>

In [38]:
A[:2,:2]

<tf.Tensor: shape=(2, 2, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20]]])>

In [39]:
A[:2,:2,:2]

<tf.Tensor: shape=(2, 2, 2), dtype=int32, numpy=
array([[[ 1,  2],
        [ 5,  6]],

       [[13, 14],
        [17, 18]]])>

In [41]:
A[:-1,:-1,:-1]

<tf.Tensor: shape=(1, 2, 3), dtype=int32, numpy=
array([[[1, 2, 3],
        [5, 6, 7]]])>

In [45]:
A[:,:,-2:]

<tf.Tensor: shape=(2, 3, 2), dtype=int32, numpy=
array([[[ 3,  4],
        [ 7,  8],
        [11, 12]],

       [[15, 16],
        [19, 20],
        [23, 24]]])>

In [3]:
r4tensor = tf.zeros([2, 3, 4, 5])
r4tensor

<tf.Tensor: shape=(2, 3, 4, 5), dtype=float32, numpy=
array([[[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]],


       [[[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]],

        [[0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0.]]]], dtype=float32)>

In [11]:
r2tensor = tf.constant([[10,7],
                       [3,4]])
r2tensor.shape

TensorShape([2, 2])

In [12]:
r2tensor.ndim

2

In [8]:
r2tensor[:,-1]

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([7, 4])>

In [13]:
r3tensor = r2tensor[...,tf.newaxis]
r3tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]])>

In [14]:
r3tensor = r2tensor[:,:,tf.newaxis]
r3tensor

<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]])>

In [15]:
tf.expand_dims(r2tensor, axis= -1)


<tf.Tensor: shape=(2, 2, 1), dtype=int32, numpy=
array([[[10],
        [ 7]],

       [[ 3],
        [ 4]]])>

In [16]:
tf.expand_dims(r2tensor, axis= 0)

<tf.Tensor: shape=(1, 2, 2), dtype=int32, numpy=
array([[[10,  7],
        [ 3,  4]]])>

In [17]:
some_list = [1,2,3,4]
some_list[:2]

[1, 2]

In [30]:
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]])>

In [35]:
A[:,:1,:2]

<tf.Tensor: shape=(2, 1, 2), dtype=int32, numpy=
array([[[ 1,  2]],

       [[13, 14]]])>

# Manipulating tensors (tensor operations)

## basic operations

In [36]:
A + 10

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[11, 12, 13, 14],
        [15, 16, 17, 18],
        [19, 20, 21, 22]],

       [[23, 24, 25, 26],
        [27, 28, 29, 30],
        [31, 32, 33, 34]]])>

In [38]:
C = A + 10

In [39]:
C

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[11, 12, 13, 14],
        [15, 16, 17, 18],
        [19, 20, 21, 22]],

       [[23, 24, 25, 26],
        [27, 28, 29, 30],
        [31, 32, 33, 34]]])>

In [40]:
A

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 1,  2,  3,  4],
        [ 5,  6,  7,  8],
        [ 9, 10, 11, 12]],

       [[13, 14, 15, 16],
        [17, 18, 19, 20],
        [21, 22, 23, 24]]])>

In [41]:
A*2

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 2,  4,  6,  8],
        [10, 12, 14, 16],
        [18, 20, 22, 24]],

       [[26, 28, 30, 32],
        [34, 36, 38, 40],
        [42, 44, 46, 48]]])>

In [42]:
type(A)

tensorflow.python.framework.ops.EagerTensor

### built-in TF function

In [45]:
C = tf.add(A,8)
C

<tf.Tensor: shape=(2, 3, 4), dtype=int32, numpy=
array([[[ 9, 10, 11, 12],
        [13, 14, 15, 16],
        [17, 18, 19, 20]],

       [[21, 22, 23, 24],
        [25, 26, 27, 28],
        [29, 30, 31, 32]]])>

In [57]:
%%timeit
C = tf.add(A,8)

137 µs ± 9.19 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [56]:
A_numpy =np.arange(1,25,1)
A_numpy = np.reshape(A_numpy,(2,3,4))
type(A_numpy)

numpy.ndarray

In [51]:
A_numpy

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]],

       [[12, 13, 14, 15],
        [16, 17, 18, 19],
        [20, 21, 22, 23]]])

In [58]:
%%timeit
C_numpy = np.add(A,8)

14.2 µs ± 1.03 µs per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


https://realpython.com/numpy-tensorflow-performance/#conclusion

While the NumPy example proved quicker by a hair than TensorFlow in this case, it’s important to note that TensorFlow really shines for more complex cases. With our relatively elementary regression problem, using TensorFlow arguably amounts to “using a sledgehammer to crack a nut,” as the saying goes.

## matrix multiplication

dot matrix not element wide
dot product
`tf.linalg.matmul`

In [4]:
tensor = tf.constant([[10,7],
                     [3,4]
                     ])

In [5]:
tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[10,  7],
       [ 3,  4]])>

In [6]:
print(tensor)

tf.Tensor(
[[10  7]
 [ 3  4]], shape=(2, 2), dtype=int32)


In [7]:
tf.matmul(tensor,tensor)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]])>

### matrix multiplication with python operator "@"

In [8]:
tensor @ tensor   # dot product

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[121,  98],
       [ 42,  37]])>

In [9]:
tensor * tensor

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[100,  49],
       [  9,  16]])>

In [10]:
A = tf.constant([[1,2,5],
                [7,2,1],
                [3,3,3]
                ])
A

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[1, 2, 5],
       [7, 2, 1],
       [3, 3, 3]])>

In [11]:
B = tf.constant([[3,5],
                [6,7],
                [1,8]
                ])
B

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[3, 5],
       [6, 7],
       [1, 8]])>

In [12]:
A @ B

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[20, 59],
       [34, 57],
       [30, 60]])>

In [13]:
tf.matmul(A,B)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[20, 59],
       [34, 57],
       [30, 60]])>

In [14]:
tf.linalg.matmul(A,B)

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[20, 59],
       [34, 57],
       [30, 60]])>

In [15]:
# A * B   >>>> # error : size incompatible

In [16]:
A.shape, B.shape

(TensorShape([3, 3]), TensorShape([3, 2]))

In [17]:
X = tf.constant([[1,2],
                [3,4],
                [5,6]
                ])

In [18]:
Y = tf.constant([[7,8],
                [9,10],
                [11,12]
                ])

In [19]:
X * Y

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 7, 16],
       [27, 40],
       [55, 72]])>

In [20]:
# X @ Y   # error: size incompatible

In [21]:
# tf.matmul(X,Y) # error: size incompatible

In [22]:
Y_new = tf.reshape(Y,shape =(2,3))
Y_new

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  8,  9],
       [10, 11, 12]])>

In [23]:
X @ Y_new

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 27,  30,  33],
       [ 61,  68,  75],
       [ 95, 106, 117]])>

In [24]:
Y_new2 = tf.transpose(Y)
Y_new2

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  9, 11],
       [ 8, 10, 12]])>

In [28]:
tf.tensordot(X,Y_new,axes=1)

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 27,  30,  33],
       [ 61,  68,  75],
       [ 95, 106, 117]])>

In [29]:
tf.tensordot(X,Y_new,axes=0)

<tf.Tensor: shape=(3, 2, 2, 3), dtype=int32, numpy=
array([[[[ 7,  8,  9],
         [10, 11, 12]],

        [[14, 16, 18],
         [20, 22, 24]]],


       [[[21, 24, 27],
         [30, 33, 36]],

        [[28, 32, 36],
         [40, 44, 48]]],


       [[[35, 40, 45],
         [50, 55, 60]],

        [[42, 48, 54],
         [60, 66, 72]]]])>

https://www.tensorflow.org/api_docs/python/tf/tensordot
If axes=0, computes the outer product between a and b.

In [30]:
tf.matmul(X, tf.transpose(Y))

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 23,  29,  35],
       [ 53,  67,  81],
       [ 83, 105, 127]])>

In [32]:
tf.matmul(X,tf.reshape(Y,(2,-1)))  # X

<tf.Tensor: shape=(3, 3), dtype=int32, numpy=
array([[ 27,  30,  33],
       [ 61,  68,  75],
       [ 95, 106, 117]])>

In [35]:
Y

<tf.Tensor: shape=(3, 2), dtype=int32, numpy=
array([[ 7,  8],
       [ 9, 10],
       [11, 12]])>

In [33]:
Y_trans = tf.transpose(Y)
Y_trans

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  9, 11],
       [ 8, 10, 12]])>

In [34]:
Y_reshaped = tf.reshape(Y, (2,-1))
Y_reshaped

<tf.Tensor: shape=(2, 3), dtype=int32, numpy=
array([[ 7,  8,  9],
       [10, 11, 12]])>

In [36]:
Q = tf.constant([[1,2],
                [3,4],
                ])
Q

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[1, 2],
       [3, 4]])>

In [37]:
R = tf.constant([[5,6],
                [7,8],
                ])
R

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[5, 6],
       [7, 8]])>

In [38]:
Q @ R

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[19, 22],
       [43, 50]])>

In [39]:
tf.matmul(Q,R)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[19, 22],
       [43, 50]])>

In [40]:
tf.tensordot(Q,R,axes=1)

<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
array([[19, 22],
       [43, 50]])>

In [41]:
tf.tensordot(Q,R,axes=0)

<tf.Tensor: shape=(2, 2, 2, 2), dtype=int32, numpy=
array([[[[ 5,  6],
         [ 7,  8]],

        [[10, 12],
         [14, 16]]],


       [[[15, 18],
         [21, 24]],

        [[20, 24],
         [28, 32]]]])>

# changing datatype of tensor

reduce memory in RAM, but reduce calculation accuracy

In [65]:
B = tf.constant([1.7,7.4])
B

<tf.Tensor: shape=(2,), dtype=float32, numpy=array([1.7, 7.4], dtype=float32)>

In [66]:
C = tf.constant([7,10])
C

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10])>

In [67]:
D = tf.cast(B, dtype=tf.float16)
D

<tf.Tensor: shape=(2,), dtype=float16, numpy=array([1.7, 7.4], dtype=float16)>

# tensor aggregation

In [4]:
D = tf.constant([-7,-10])
D

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ -7, -10])>

In [5]:
tf.abs(D)

<tf.Tensor: shape=(2,), dtype=int32, numpy=array([ 7, 10])>

### aggregation
- minimum
- maximum
- mean
- sum

In [6]:
A = tf.constant(np.random.randint(0,100,size=50))
A

<tf.Tensor: shape=(50,), dtype=int32, numpy=
array([68, 43, 39, 35,  4, 16, 29, 99, 69, 46, 99, 58,  4, 42, 38, 81, 19,
       28, 98, 31, 39, 39, 99,  2, 28, 37, 96, 97,  0, 52, 72, 33,  6, 89,
       48, 90, 13,  9, 65, 76, 85,  7, 28, 30,  7, 54, 75, 27,  8, 78])>

In [7]:
tf.size(A)

<tf.Tensor: shape=(), dtype=int32, numpy=50>

In [8]:
B =tf.reshape(A,(5,-1))
B

<tf.Tensor: shape=(5, 10), dtype=int32, numpy=
array([[68, 43, 39, 35,  4, 16, 29, 99, 69, 46],
       [99, 58,  4, 42, 38, 81, 19, 28, 98, 31],
       [39, 39, 99,  2, 28, 37, 96, 97,  0, 52],
       [72, 33,  6, 89, 48, 90, 13,  9, 65, 76],
       [85,  7, 28, 30,  7, 54, 75, 27,  8, 78]])>

In [9]:
tf.reduce_max(B), tf.reduce_min(B), tf.reduce_mean(B)

(<tf.Tensor: shape=(), dtype=int32, numpy=99>,
 <tf.Tensor: shape=(), dtype=int32, numpy=0>,
 <tf.Tensor: shape=(), dtype=int32, numpy=46>)

In [10]:
tf.math.reduce_mean(B)

<tf.Tensor: shape=(), dtype=int32, numpy=46>

*** tf cannot return std due to unfit of original dtype

In [11]:
B = tf.cast(B, dtype=tf.float32)

In [12]:
tf.math.reduce_std(B).dtype

tf.float32

In [13]:
tf.math.reduce_std(B)

<tf.Tensor: shape=(), dtype=float32, numpy=31.069117>

In [14]:
tf.math.reduce_std(B).numpy()

31.069117

https://www.tensorflow.org/probability/install

Note: Since TensorFlow is not included as a dependency of the TensorFlow Probability package (in setup.py), you must explicitly install the TensorFlow package (tensorflow or tensorflow-gpu). This allows us to maintain one package instead of separate packages for CPU and GPU-enabled TensorFlow.

In [15]:
# !pip install tensorflow-probability==0.18.0

ImportError: This version of TensorFlow Probability requires TensorFlow version >= 2.16; Detected an installation of version 2.10.1. Please upgrade TensorFlow to proceed.

This is the 0.18.0 release of TensorFlow Probability. It is tested and stable against TensorFlow version 2.10 and JAX 0.3.17 .
https://github.com/tensorflow/probability/releases

In [16]:
import tensorflow_probability as tfp

In [29]:
B_int =tf.reshape(A,(5,-1))
B_int

<tf.Tensor: shape=(5, 10), dtype=int32, numpy=
array([[68, 43, 39, 35,  4, 16, 29, 99, 69, 46],
       [99, 58,  4, 42, 38, 81, 19, 28, 98, 31],
       [39, 39, 99,  2, 28, 37, 96, 97,  0, 52],
       [72, 33,  6, 89, 48, 90, 13,  9, 65, 76],
       [85,  7, 28, 30,  7, 54, 75, 27,  8, 78]])>

In [18]:
tfp.stats.variance(B_int)

<tf.Tensor: shape=(10,), dtype=int32, numpy=array([ 400,  278, 1192,  795,  294,  749, 1089, 1456, 1426,  325])>

In [19]:
tfp.stats.variance(B)

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([ 400.23996,  278.4    , 1192.5599 ,  794.64   ,  294.4    ,
        749.04004, 1089.4401 , 1456.8    , 1426.8    ,  324.63998],
      dtype=float32)>

In [21]:
tf.math.reduce_variance(B)

<tf.Tensor: shape=(), dtype=float32, numpy=965.29>

In [24]:
tfp.stats.variance(A)

<tf.Tensor: shape=(), dtype=int32, numpy=965>

In [28]:
tf.math.reduce_variance(tf.cast(A, dtype=tf.float32))

<tf.Tensor: shape=(), dtype=float32, numpy=965.29>

## find positional max or min

In [30]:
tf.random.set_seed(42)
F = tf.random.uniform(shape=[50])
F

<tf.Tensor: shape=(50,), dtype=float32, numpy=
array([0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
       0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
       0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
       0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
       0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
       0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
       0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
       0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
       0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
       0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043],
      dtype=float32)>

### find position max 

In [32]:
tf.argmax(F)

<tf.Tensor: shape=(), dtype=int64, numpy=42>

In [35]:
F[tf.argmax(F)]

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

In [37]:
tf.reduce_max(F)

<tf.Tensor: shape=(), dtype=float32, numpy=0.9671384>

#### check equality

no error: equal

In [39]:
assert F[tf.argmax(F)] == tf.reduce_max(F)

In [40]:
F[tf.argmax(F)] == tf.reduce_max(F)

<tf.Tensor: shape=(), dtype=bool, numpy=True>

In [55]:
tf.math.argmax(F,axis=0)

<tf.Tensor: shape=(), dtype=int64, numpy=42>

In [52]:
tf.math.argmax(F,axis=-1)

<tf.Tensor: shape=(), dtype=int64, numpy=42>

In [60]:
G = tf.reshape(F,(5,-1))

In [61]:
tf.math.argmax(G,axis=0)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([3, 4, 4, 2, 4, 1, 3, 0, 2, 3], dtype=int64)>

In [62]:
tf.math.argmax(G,axis=-1)

<tf.Tensor: shape=(5,), dtype=int64, numpy=array([7, 5, 8, 0, 2], dtype=int64)>

In [63]:
tf.argmax(G)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([3, 4, 4, 2, 4, 1, 3, 0, 2, 3], dtype=int64)>

In [64]:
tf.argmax(G,axis=-1)

<tf.Tensor: shape=(5,), dtype=int64, numpy=array([7, 5, 8, 0, 2], dtype=int64)>

In [67]:
tf.argmax(G,axis=0,)

<tf.Tensor: shape=(10,), dtype=int64, numpy=array([3, 4, 4, 2, 4, 1, 3, 0, 2, 3], dtype=int64)>

Squeeze a tensor(removing all single dimensions)

In [75]:
tf.random.set_seed(42)
G = tf.constant(tf.random.uniform(shape=[50]), shape=(1,1,1,1,50))
G

<tf.Tensor: shape=(1, 1, 1, 1, 50), dtype=float32, numpy=
array([[[[[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
           0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
           0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
           0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
           0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
           0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
           0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
           0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
           0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
           0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]]]]],
      dtype=float32)>

In [82]:
tf.random.set_seed(42)
H = tf.constant(tf.random.uniform(shape=[50]), shape=(1,2,1,1,25))
H

<tf.Tensor: shape=(1, 2, 1, 1, 25), dtype=float32, numpy=
array([[[[[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
           0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
           0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
           0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
           0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ]]],


        [[[0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
           0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
           0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
           0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
           0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]]]]],
      dtype=float32)>

In [93]:
tf.random.set_seed(42)
I = tf.constant(tf.random.uniform(shape=[50]), shape=(1,2,25,1,1))
I

<tf.Tensor: shape=(1, 2, 25, 1, 1), dtype=float32, numpy=
array([[[[[0.6645621 ]],

         [[0.44100678]],

         [[0.3528825 ]],

         [[0.46448255]],

         [[0.03366041]],

         [[0.68467236]],

         [[0.74011743]],

         [[0.8724445 ]],

         [[0.22632635]],

         [[0.22319686]],

         [[0.3103881 ]],

         [[0.7223358 ]],

         [[0.13318717]],

         [[0.5480639 ]],

         [[0.5746088 ]],

         [[0.8996835 ]],

         [[0.00946367]],

         [[0.5212307 ]],

         [[0.6345445 ]],

         [[0.1993283 ]],

         [[0.72942245]],

         [[0.54583454]],

         [[0.10756552]],

         [[0.6767061 ]],

         [[0.6602763 ]]],


        [[[0.33695042]],

         [[0.60141766]],

         [[0.21062577]],

         [[0.8527372 ]],

         [[0.44062173]],

         [[0.9485276 ]],

         [[0.23752594]],

         [[0.81179297]],

         [[0.5263394 ]],

         [[0.494308  ]],

         [[0.21612847]],

    

In [94]:
G.shape

TensorShape([1, 1, 1, 1, 50])

In [95]:
G_squeezed = tf.squeeze(G)
G_squeezed.shape

TensorShape([50])

In [96]:
H_squeezed = tf.squeeze(H)
H_squeezed.shape

TensorShape([2, 25])

In [97]:
I_squeezed = tf.squeeze(I)
I_squeezed.shape

TensorShape([2, 25])

In [98]:
G_squeezed

<tf.Tensor: shape=(50,), dtype=float32, numpy=
array([0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
       0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
       0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
       0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
       0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ,
       0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
       0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
       0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
       0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
       0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043],
      dtype=float32)>

In [106]:
H_squeezed

<tf.Tensor: shape=(2, 25), dtype=float32, numpy=
array([[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
        0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
        0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
        0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
        0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ],
       [0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
        0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
        0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
        0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
        0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]],
      dtype=float32)>

In [107]:
I_squeezed

<tf.Tensor: shape=(2, 25), dtype=float32, numpy=
array([[0.6645621 , 0.44100678, 0.3528825 , 0.46448255, 0.03366041,
        0.68467236, 0.74011743, 0.8724445 , 0.22632635, 0.22319686,
        0.3103881 , 0.7223358 , 0.13318717, 0.5480639 , 0.5746088 ,
        0.8996835 , 0.00946367, 0.5212307 , 0.6345445 , 0.1993283 ,
        0.72942245, 0.54583454, 0.10756552, 0.6767061 , 0.6602763 ],
       [0.33695042, 0.60141766, 0.21062577, 0.8527372 , 0.44062173,
        0.9485276 , 0.23752594, 0.81179297, 0.5263394 , 0.494308  ,
        0.21612847, 0.8457197 , 0.8718841 , 0.3083862 , 0.6868038 ,
        0.23764038, 0.7817228 , 0.9671384 , 0.06870162, 0.79873943,
        0.66028714, 0.5871513 , 0.16461694, 0.7381023 , 0.32054043]],
      dtype=float32)>

In [108]:
H_squeezed == I_squeezed

<tf.Tensor: shape=(2, 25), dtype=bool, numpy=
array([[ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True]])>

In [119]:
tf.math.equal(H_squeezed, I_squeezed)

<tf.Tensor: shape=(2, 25), dtype=bool, numpy=
array([[ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True],
       [ True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True,  True,  True,  True,  True]])>

## one-hot encoding

In [42]:
list = [0,1,2,3]  # 4 level e.g. red, green, blue, purple

In [43]:
tf.one_hot(list, depth=4)

<tf.Tensor: shape=(4, 4), dtype=float32, numpy=
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]], dtype=float32)>

In [46]:
tf.one_hot(list, depth=4, on_value="on", off_value="off")

<tf.Tensor: shape=(4, 4), dtype=string, numpy=
array([[b'on', b'off', b'off', b'off'],
       [b'off', b'on', b'off', b'off'],
       [b'off', b'off', b'on', b'off'],
       [b'off', b'off', b'off', b'on']], dtype=object)>

In [63]:
indices = [[0, 2], [1, -1]]
depth = 3
tf.one_hot(indices, depth)

<tf.Tensor: shape=(2, 2, 3), dtype=float32, numpy=
array([[[1., 0., 0.],
        [0., 0., 1.]],

       [[0., 1., 0.],
        [0., 0., 0.]]], dtype=float32)>

In [64]:
type(indices)

list

In [55]:
Q, Q.shape

(<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
 array([[1, 2],
        [3, 4]])>,
 TensorShape([2, 2]))

In [54]:
tf.one_hot(Q, depth=4)

<tf.Tensor: shape=(2, 2, 4), dtype=float32, numpy=
array([[[0., 1., 0., 0.],
        [0., 0., 1., 0.]],

       [[0., 0., 0., 1.],
        [0., 0., 0., 0.]]], dtype=float32)>

## square, sqrt, log

In [69]:
H = tf.range(0,10)
H

<tf.Tensor: shape=(10,), dtype=int32, numpy=array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])>

In [70]:
tf.square(H)

<tf.Tensor: shape=(10,), dtype=int32, numpy=array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])>

In [77]:
K = tf.sqrt(tf.cast(H,tf.float32))
K

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([0.       , 1.       , 1.4142135, 1.7320508, 2.       , 2.2360678,
       2.4494896, 2.6457512, 2.828427 , 3.       ], dtype=float32)>

In [78]:
L = tf.math.log(tf.cast(H,tf.float32))
L

<tf.Tensor: shape=(10,), dtype=float32, numpy=
array([     -inf, 0.       , 0.6931472, 1.0986123, 1.3862944, 1.609438 ,
       1.7917595, 1.9459102, 2.0794415, 2.1972246], dtype=float32)>

## Tensors and Numpy

In [80]:
J = tf.constant(np.array([3.,7.,10.]))
J

<tf.Tensor: shape=(3,), dtype=float64, numpy=array([ 3.,  7., 10.])>

In [82]:
np.array(J), type(np.array(J))

(array([ 3.,  7., 10.]), numpy.ndarray)

In [83]:
J.numpy(), type(J.numpy())

(array([ 3.,  7., 10.]), numpy.ndarray)

In [85]:
J = tf.constant([3.])
J.numpy()[0]

3.0

In [88]:
Q, Q.numpy()[1][1]

(<tf.Tensor: shape=(2, 2), dtype=int32, numpy=
 array([[1, 2],
        [3, 4]])>,
 4)

In [94]:
# default type of numpy and tensor
numpy_J = tf.constant(np.array([3.,7.,10.]))
tensor_J = tf.constant([3.,7.,10.])
numpy_J.dtype, tensor_J.dtype

(tf.float64, tf.float32)

# check GPU

In [98]:
tf.test.is_gpu_available()

True

In [101]:
tf.config.list_physical_devices()

[PhysicalDevice(name='/physical_device:CPU:0', device_type='CPU'),
 PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [107]:
tf.config.list_physical_devices('GPU')

[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]

In [108]:
!nvidia-smi

Thu Jun 13 11:39:54 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 552.44                 Driver Version: 552.44         CUDA Version: 12.4     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                     TCC/WDDM  | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3050 ...  WDDM  |   00000000:01:00.0 Off |                  N/A |
| N/A   42C    P8              5W /   20W |    1837MiB /   4096MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [6]:
X = tf.range(start=1,limit=10,delta=1)
X

<tf.Tensor: shape=(9,), dtype=int32, numpy=array([1, 2, 3, 4, 5, 6, 7, 8, 9])>

In [7]:
X = tf.range(1,10,delta=1)
X

<tf.Tensor: shape=(9,), dtype=int32, numpy=array([1, 2, 3, 4, 5, 6, 7, 8, 9])>

In [8]:
X = tf.range(1,10,delta=2)
X

<tf.Tensor: shape=(5,), dtype=int32, numpy=array([1, 3, 5, 7, 9])>

In [10]:
X = tf.cast(X, dtype=tf.float32)
# tf.float(16,32,64), tf.int(8,16,32,64)